lib/model/internal_link.dart

```diff
@@ -0,0 +1,282 @@
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final fragment = StringBuffer('narrow');
+  for (ApiNarrowElement element in apiNarrow) {
+    fragment.write('/');
+    if (element.negated) {
+      fragment.write('-');
+    }
+
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      element = element.resolve(legacy: !supportsOperatorDm);
+    }
+
+    fragment.write('${element.operator}/');
+
+    switch (element) {
+      case ApiNarrowStream():
+        final streamId = element.operand;
+        final name = store.streams[streamId]?.name ?? 'unknown';
+        final slugifiedName = _encodeHashComponent(name.replaceAll(' ', '-'));
+        fragment.write('$streamId-$slugifiedName');
+      case ApiNarrowTopic():
+        fragment.write(_encodeHashComponent(element.operand));
+      case ApiNarrowDmModern():
+        final suffix = element.operand.length >= 3 ? 'group' : 'dm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowPmWith():
+        final suffix = element.operand.length >= 3 ? 'group' : 'pm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowDm():
+        assert(false, 'ApiNarrowDm should have been resolved');
+      case ApiNarrowMessageId():
+        fragment.write(element.operand.toString());
+    }
+  }
+
+  if (nearMessageId != null) {
+    fragment.write('/near/$nearMessageId');
+  }
+
+  return store.account.realmUrl.replace(fragment: fragment.toString());
+}
+
+/// A [Narrow] from a given URL, on `store`'s realm.
+///
+/// Returns `null` if any of the operator/operand pairs are invalid.
+///
+/// Since narrow links can combine operators in ways our [Narrow] type can't
+/// represent, this can also return null for valid narrow links.
+///
+/// This can also return null for some valid narrow links that our Narrow
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+Narrow? parseInternalLink(Uri url, PerAccountStore store) {
+  if (!_isInternalLink(url, store.account.realmUrl)) return null;
+
+  final (category, segments) = _getCategoryAndSegmentsFromFragment(url.fragment);
+  switch (category) {
+    case 'narrow':
+      if (segments.isEmpty || !segments.length.isEven) return null;
+      return _interpretNarrowSegments(segments, store);
+  }
+  return null;
+}
+
+/// Check if `url` is an internal link on the given `realmUrl`.
+bool _isInternalLink(Uri url, Uri realmUrl) {
+  if (url.hasScheme || url.hasAuthority || url.hasPort) {
+    // input that do not include scheme, host, and port
+    // are treated as on realm
```

Comment:
It really worries me to be adding ad-hoc logic like this in the context of interpreting a URL. That's the sort of thing that creates security vulnerabilities: one system thinks the URL means one thing and refers to something trusted, but then passes the URL to a different system which interprets it as pointing to something else that in fact should not be trusted.

So any time we think there's a need for this sort of logic, we need to be very explicit about thinking through what we're actually trying to do, and why the logic we're writing will correctly do the thing we want and not have edge cases where it goes wrong.

In this instance, we're porting over code from zulip-mobile, and I don't think there's corresponding logic there. So I suspect we ultimately don't need this.

If I take this condition out, I see there are a couple of tests that fail, which I assume is related to why you added it. (And https://github.com/zulip/zulip-flutter/pull/305#discussion_r1334363539 sounds related too.) I recommend starting a chat thread in `#mobile-team` with your observations and debugging of what goes wrong without this condition; that'll be a better medium than GitHub for a debugging or "what's the right way to solve this problem" conversation.

Reply:
Through our discussion we settled on running all input through `realmUrl.resolve()` so that they have an origin to report later. This matches the mobile behavior of running input through `new URL()` with a base.

Reply:
Linking that chat discussion for cross-reference: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/URLs.20are.20tricky.20to.20parse/near/1647599